<h1 style="font-size: 36px; text-align: center !important">Projet IA [Human For You]</h1>


# Livrable : 

## Sous-titre: Livrable Technique - Présentation

---

### GROUP 04 
- LAGGOUNE Amine
- SACLIER Victor
- BOISSEL Arnaud



---

<center>
<img src='https://blog.integral-system.fr/wp-content/uploads/2020/02/33797061-1920x1359.jpg' />
</center>

---

### Date: [31/01/2024]



## Introduction



---

## Objectifs

- Préparation de notre jeux de données.


- Etude des différents modèles d'apprentissage.


- Le choix du modèle retenu (Avec justification).


- Nos résultats finaux avec proposition de pistes d'amélioration.

---

## Table des Matières

1. Chapitre 1
2. Chapitre 2
3. Chapitre 3
4. Conclusion

---







---

<center>
© [FISA  A5 INFO], [2024]
</center>


In [59]:
# imports
import os
import numpy as np
from numpy.random import default_rng
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import hvplot.pandas
import seaborn as sn
import warnings
from sklearn.impute import SimpleImputer

# on filtre les warnings
warnings.filterwarnings('ignore')

# stabilité du notebook d'une exécution à l'autre
random = default_rng(42)

# Ajustement des tailles des labels graphiques utilisé dans le notebook
plt.rcParams['axes.labelsize'] = 14
plt.rcParams['xtick.labelsize'] = 12
plt.rcParams['ytick.labelsize'] = 12

Implementation des fonctions pour Importer, Lire et exporter les données de notre dataset

In [60]:
PROJECT_ROOT_DIR = "."
DATA_PATH = "datasets"
IMAGES_PATH = os.path.join(PROJECT_ROOT_DIR, "images") # le dossier doit exister

def save_fig(fig_id, tight_layout=True, fig_extension="png", resolution=300):
    path = os.path.join(IMAGES_PATH, fig_id + "." + fig_extension)
    print("Saving figure", fig_id)
    if tight_layout:
        plt.tight_layout()
    plt.savefig(path, format=fig_extension, dpi=resolution)


def load_data(path, filename):
    file_path = os.path.join(path, filename)
    return pd.read_csv(file_path)

### Import des données issue des datasets
- __employee_data__ : Cette variable va contenir les informations générales des employées.
- __employee_survey__ : Cette variable va contenir les informations d'enquêtes des employées.
- __manager_survey__ : Cette variable va contenir les informations d'enquêtes des managers.
- __in_time__ et __out_time__ : Ces deux variables vont contenir les données du pointage des employées, càd, le temps d'entrée et de sortie de chaque employée.

In [61]:
employee_data = load_data(DATA_PATH, 'general_data.csv')
employee_survey = load_data(DATA_PATH, 'employee_survey_data.csv')
manager_survey = load_data(DATA_PATH, 'manager_survey_data.csv')
in_time = load_data(DATA_PATH, 'in_time.csv')
out_time = load_data(DATA_PATH, 'out_time.csv')

# Analyse préliminaire
##  1- Analyse de données des horaires de travail

In [62]:
in_time.head(4)


Unnamed: 0  2015-01-01           2015-01-02           2015-01-05  \
0           1         NaN  2015-01-02 09:43:45  2015-01-05 10:08:48   
1           2         NaN  2015-01-02 10:15:44  2015-01-05 10:21:05   
2           3         NaN  2015-01-02 10:17:41  2015-01-05 09:50:50   
3           4         NaN  2015-01-02 10:05:06  2015-01-05 09:56:32   

            2015-01-06           2015-01-07           2015-01-08  \
0  2015-01-06 09:54:26  2015-01-07 09:34:31  2015-01-08 09:51:09   
1                  NaN  2015-01-07 09:45:17  2015-01-08 10:09:04   
2  2015-01-06 10:14:13  2015-01-07 09:47:27  2015-01-08 10:03:40   
3  2015-01-06 10:11:07  2015-01-07 09:37:30  2015-01-08 10:02:08   

            2015-01-09           2015-01-12           2015-01-13  2015-01-14  \
0  2015-01-09 10:09:25  2015-01-12 09:42:53  2015-01-13 10:13:06         NaN   
1  2015-01-09 09:43:26  2015-01-12 10:00:07  2015-01-13 10:43:29         NaN   
2  2015-01-09 10:05:49  2015-01-12 10:03:47  2015-01-13 10:21:26         NaN   
3  2015-01-09 10:08:12  2015-01-12 10:13:42  2015-01-13 09:53:22         NaN   

            2015-01-15           2015-01-16           2015-01-19  \
0  2015-01-15 10:01:24  2015-01-16 10:19:08                  NaN   
1  2015-01-15 09:37:57  2015-01-16 09:57:18  2015-01-19 10:23:43   
2  2015-01-15 09:55:11  2015-01-16 10:05:36  2015-01-19 09:47:53   
3  2015-01-15 10:00:50  2015-01-16 09:58:06  2015-01-19 09:43:11   

            2015-01-20  ...           2015-12-11           2015-12-14  \
0  2015-01-20 09:50:34  ...  2015-12-11 09:46:49  2015-12-14 10:03:33   
1  2015-01-20 09:29:03  ...  2015-12-11 10:15:18  2015-12-14 10:00:01   
2  2015-01-20 09:57:09  ...  2015-12-11 09:49:55  2015-12-14 09:42:31   
3  2015-01-20 10:29:06  ...  2015-12-11 09:46:02  2015-12-14 10:20:36   

            2015-12-15           2015-12-16           2015-12-17  \
0                  NaN  2015-12-16 10:21:19                  NaN   
1  2015-12-15 10:30:18  2015-12-16 10:01:18  2015-12-17 09:15:08   
2  2015-12-15 09:54:48  2015-12-16 09:48:34  2015-12-17 09:53:17   
3  2015-12-15 09:37:17  2015-12-16 10:22:09  2015-12-17 09:54:36   

            2015-12-18           2015-12-21           2015-12-22  \
0                  NaN  2015-12-21 09:55:29  2015-12-22 10:04:06   
1  2015-12-18 10:37:17  2015-12-21 09:49:02  2015-12-22 10:33:51   
2  2015-12-18 10:15:14  2015-12-21 10:10:28  2015-12-22 09:44:44   
3  2015-12-18 10:17:38  2015-12-21 09:58:21  2015-12-22 10:04:25   

            2015-12-23           2015-12-24 2015-12-25           2015-12-28  \
0  2015-12-23 10:14:27  2015-12-24 10:11:35        NaN  2015-12-28 10:13:41   
1  2015-12-23 10:12:10                  NaN        NaN  2015-12-28 09:31:45   
2  2015-12-23 10:15:54  2015-12-24 10:07:26        NaN  2015-12-28 09:42:05   
3  2015-12-23 10:11:46  2015-12-24 09:43:15        NaN  2015-12-28 09:52:44   

            2015-12-29           2015-12-30           2015-12-31  
0  2015-12-29 10:03:36  2015-12-30 09:54:12  2015-12-31 10:12:44  
1  2015-12-29 09:55:49  2015-12-30 10:32:25  2015-12-31 09:27:20  
2  2015-12-29 09:43:36  2015-12-30 09:34:05  2015-12-31 10:28:39  
3  2015-12-29 09:33:16  2015-12-30 10:18:12  2015-12-31 10:01:15  

[4 rows x 262 columns]

In [63]:
in_time.info()
out_time.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Columns: 262 entries, Unnamed: 0 to 2015-12-31
dtypes: float64(12), int64(1), object(249)
memory usage: 8.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Columns: 262 entries, Unnamed: 0 to 2015-12-31
dtypes: float64(12), int64(1), object(249)
memory usage: 8.8+ MB


### 1.1Analyse du DataFrame `in_time`

### Structure Générale
- **Nombre d'Entrées (Lignes):** 4410
- **Nombre de Colonnes:** 262
- **Type de Données:** 
  - Numériques: `float64` (12 colonnes), `int64` (1 colonne)
  - Catégorielles/Textuelles: `object` (249 colonnes)



##  2- Analyse de données issue du service des ressources humaines

In [64]:
employee_data.head(1)

Age Attrition BusinessTravel Department  DistanceFromHome  Education  \
0   51        No  Travel_Rarely      Sales                 6          2   

  EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  \
0  Life Sciences              1           1  Female         1   

                     JobRole MaritalStatus  MonthlyIncome  NumCompaniesWorked  \
0  Healthcare Representative       Married         131160                 1.0   

  Over18  PercentSalaryHike  StandardHours  StockOptionLevel  \
0      Y                 11              8                 0   

   TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                1.0                      6               1   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0

In [65]:
employee_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

### 2.1 Analyse du DataFrame `employee_data`  

### Aperçu Général
- **Entrées (Lignes):** 4410
- **Colonnes:** 24
- **Type de Données:** 
  - Numériques: `int64` et `float64`
  - Catégorielles: `object`

### Gestion des Données Manquantes
- **Colonnes avec Données Manquantes:** 
  - `NumCompaniesWorked` <span style="color:red;">(19 valeurs manquantes)</span>
  - `TotalWorkingYears` <span style="color:red;">(9 valeurs manquantes)</span>


### Détails des Colonnes
- **Variables Numériques:** 
  - Exemples: `Age`, `DistanceFromHome`, `Education`, `MonthlyIncome`
- **Variables Catégorielles:** 
  - Exemples: `Attrition`, `BusinessTravel`, `Department`
 



##  2- Analyse de données évaluation du manager &  Enquête qualité de vie au travail

In [66]:
manager_survey.head(1)

EmployeeID  JobInvolvement  PerformanceRating
0           1               3                  3

In [67]:
employee_survey.head(1)

EmployeeID  EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance
0           1                      3.0              4.0              2.0

In [68]:
manager_survey.info()
employee_survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   EmployeeID         4410 non-null   int64
 1   JobInvolvement     4410 non-null   int64
 2   PerformanceRating  4410 non-null   int64
dtypes: int64(3)
memory usage: 103.5 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 4 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   EmployeeID               4410 non-null   int64  
 1   EnvironmentSatisfaction  4385 non-null   float64
 2   JobSatisfaction          4390 non-null   float64
 3   WorkLifeBalance          4372 non-null   float64
dtypes: float64(3), int64(1)
memory usage: 137.9 KB



## Structure Générale
- **Nombre d'Entrées (Lignes):** 4410
- **Nombre de Colonnes:** 4
- **Types de Données:**
  - `int64`: 1 colonne (`EmployeeID`)
  - `float64`: 3 colonnes (`EnvironmentSatisfaction`, `JobSatisfaction`, `WorkLifeBalance`)

## Détails des Colonnes
- **`EmployeeID` (Intégralité des Données Présentes):**
  - 4410 valeurs non-nulles, identifiant unique pour chaque employé.
- **`EnvironmentSatisfaction` (25 Valeurs Manquantes):**
  - 4385 valeurs non-nulles, évaluation de la satisfaction environnementale (float).
- **`JobSatisfaction` (20 Valeurs Manquantes):**
  - 4390 valeurs non-nulles, évaluation de la satisfaction professionnelle (float).
- **`WorkLifeBalance` (38 Valeurs Manquantes):**
  - 4372 valeurs non-nulles, évaluation de l'équilibre travail-vie personnelle (float).



# Conclusion de l'Analyse Préliminaire

## Observations Clés
- Les DataFrames `in_time` et `out_time` **ne contiennent pas de colonne `EmployeeID`**. Cependant, ils incluent une colonne `Unnamed` qui, après examen, semble faire référence à l'`EmployeeID`.

- **Présence de Valeurs 'NaN'** : Ces valeurs sont observées dans certaines variables et pourraient être interprétées comme des jours où les employés étaient absents.

## Prochaines Étapes
- **Calcul du Temps Moyen Passé en Entreprise** : L'objectif immédiat est de calculer la durée moyenne de présence des employés sur le lieu de travail.

- **Fusion des Données** : Après avoir calculé cette métrique, nous prévoyons de fusionner `in_time`, `out_time` et d'autres informations pertinentes dans **un seul et unique DataFrame**. Ce processus facilitera une analyse globale et détaillée des données des employés.



On fait une copie de `in_time` et `out_time`  pour eviter de modifier les originaux 

In [69]:
InTime_copy = in_time.copy()
OutTime_copy = out_time.copy()

In [70]:
InTime_copy.rename(columns={InTime_copy.columns[0]: 'EmployeeID'}, inplace=True)
OutTime_copy.rename(columns={OutTime_copy.columns[0]: 'EmployeeID'}, inplace=True)

In [71]:
for date in InTime_copy.columns[1:]:  # en supposant que la première colonne est EmployeeID
    InTime_copy[date] = pd.to_datetime(InTime_copy[date])
    OutTime_copy[date] = pd.to_datetime(OutTime_copy[date])

    # Remplacer NaN par 0 avant de calculer la différence

    InTime_copy[date] = (OutTime_copy[date] - InTime_copy[date]).dt.total_seconds() / 3600


In [74]:
InTime_copy.head()

EmployeeID  2015-01-01  2015-01-02  2015-01-05  2015-01-06  2015-01-07  \
0           1         NaN    7.208333    7.189722    7.410833    7.006667   
1           2         NaN    8.109167    7.454722         NaN    7.396944   
2           3         NaN    6.692500    7.265556    6.405278    6.765000   
3           4         NaN    7.338333    7.291944    6.943056    6.919444   
4           5         NaN    8.055556    7.988056    7.682500    7.806111   

   2015-01-08  2015-01-09  2015-01-12  2015-01-13  2015-01-14  2015-01-15  \
0    7.289722    7.484444    7.262778    7.831111         NaN    7.346944   
1    7.416667    7.150833    7.611389    7.278889         NaN    7.613056   
2    7.345000    6.861389    7.418611    6.999722         NaN    7.438333   
3    6.850833    7.193056    6.998611    7.306389         NaN    6.876667   
4    7.662222    7.721667    8.365000    8.257222         NaN    8.260000   

   2015-01-16  2015-01-19  2015-01-20  ...  2015-12-11  2015-12-14  \
0    7.267500         NaN    6.775833  ...    7.651944    7.538889   
1    7.727500    7.577500    7.602778  ...    7.458056    7.536111   
2    7.210278    7.072222    6.920556  ...    7.199444    7.096389   
3    6.907778    6.518611    7.178889  ...    6.966667    7.055833   
4    8.403611         NaN    7.815278  ...    8.695833    7.589444   

   2015-12-15  2015-12-16  2015-12-17  2015-12-18  2015-12-21  2015-12-22  \
0         NaN    7.551944         NaN         NaN    7.339167    7.395833   
1    7.435278    8.255278    7.928333    7.903056    7.753889    7.712222   
2    6.751111    6.902778    7.218333    6.785833    7.163611    6.801667   
3    6.999444    7.313056    7.455833    7.629167    6.846667    7.326389   
4    8.193056    8.129444    8.320000    7.903611    7.665000    7.957500   

   2015-12-23  2015-12-24  2015-12-25  2015-12-28  2015-12-29  2015-12-30  \
0    6.504722    7.596389         NaN    7.773889    7.315000    7.778889   
1    7.435556         NaN         NaN    7.614722    7.982500    7.986111   
2    6.730278    6.849722         NaN    7.023889    7.438889    7.538889   
3    7.413611    7.085000         NaN    7.447222    7.416667    7.366389   
4    7.786944    8.249444         NaN    7.662222    8.268611    7.953333   

   2015-12-31  
0    7.080278  
1    8.227222  
2    6.786389  
3    7.133056  
4    8.018056  

[5 rows x 262 columns]

In [75]:
# Calcul de la moyenne
average_presence = InTime_copy.iloc[:, 1:].mean(axis=1)
average_presence_df = pd.DataFrame({'EmployeeID': InTime_copy['EmployeeID'], 'AveragePresence': average_presence})
average_presence_df

EmployeeID  AveragePresence
0              1         7.373651
1              2         7.718969
2              3         7.013240
3              4         7.193678
4              5         8.006175
...          ...              ...
4405        4406         8.522277
4406        4407         6.092954
4407        4408         7.706632
4408        4409         9.492595
4409        4410         6.988948

[4410 rows x 2 columns]

In [ ]:
# Fusion avec general_data
merged_data_with_time = employee_data.merge(average_presence_df, on='EmployeeID')
merged_data_with_survey = merged_data_with_time.merge(manager_survey, on='EmployeeID')
general_dataset =  merged_data_with_survey.merge(employee_survey,on='EmployeeID')

In [ ]:
general_dataset.head(5)

In [ ]:
general_dataset.info()
general_dataset.describe()

In [ ]:
unique_job_roles = general_dataset['EducationField'].unique()

In [ ]:
unique_job_roles

In [ ]:
general_data_copy = general_dataset.copy()
compatibility_dict = {
    "Life Sciences": ["Healthcare Representative", "Laboratory Technician", "Research Scientist", "Research Director"],
    "Medical": ["Healthcare Representative", "Laboratory Technician", "Research Scientist", "Research Director"],
    "Marketing": ["Sales Executive", "Sales Representative"],
    "Technical Degree": ["Research Scientist", "Manufacturing Director", "Laboratory Technician"],
    "Human Resources": ["Human Resources"],
    "Other": ["Sales Executive", "Human Resources", "Research Scientist", "Manager"],
    
}
def check_compatibility(row):
    compatible_roles = compatibility_dict.get(row['EducationField'], [])
    if row['JobRole'] in compatible_roles:
        return 1
    return 0



general_data_copy['JobCompatibility'] = general_data_copy.apply(check_compatibility, axis=1)

In [ ]:
pd.set_option('display.max_columns', 31)
general_data_copy['Attrition'] = general_data_copy['Attrition'].apply(lambda x: 1 if x == 'Yes' else 0)
general_data_copy.head()

In [ ]:
general_data_copy.hist(figsize=(40, 40))

In [ ]:
import matplotlib.pyplot as plt
import os
import pandas as pd

# Supposons que general_data_copy est votre DataFrame
# general_data_copy = ...

# Création du dossier pour les graphiques s'il n'existe pas
folder_name = 'dataset_histograms'
os.makedirs(folder_name, exist_ok=True)

# Séparation des colonnes en numériques et non numériques
numeric_columns = general_data_copy.select_dtypes(include=[np.number]).columns
non_numeric_columns = general_data_copy.select_dtypes(exclude=[np.number]).columns

# Réglages de style pour les graphiques
plt.style.use('seaborn-darkgrid')
plt.rcParams.update({'font.size': 12, 'figure.facecolor': 'white'})

# Génération et sauvegarde des histogrammes pour les colonnes numériques
for column in numeric_columns:
    plt.figure(figsize=(10, 6))
    # Les couleurs par défaut seront utilisées ici
    general_data_copy[column].hist(bins=20, edgecolor='black')
    plt.title(column)
    plt.xlabel(column)
    plt.ylabel('Frequency')
    
    # Ajout d'un quadrillage pour une meilleure lisibilité
    plt.grid(axis='y', linestyle='--', linewidth=0.5)

    # Sauvegarde de l'histogramme
    plt.savefig(f'{folder_name}/hist_{column}.png', bbox_inches='tight', facecolor='white')
    plt.close()

# Génération et sauvegarde des graphiques à barres pour les colonnes non numériques
for column in non_numeric_columns:
    value_counts = general_data_copy[column].value_counts()
    plt.figure(figsize=(10, 6))
    # Les couleurs par défaut seront utilisées ici
    value_counts.plot(kind='bar', edgecolor='black')
    plt.title(column)
    plt.xlabel(column)
    plt.ylabel('Frequency')

    # Ajout d'un quadrillage pour une meilleure lisibilité
    plt.grid(axis='y', linestyle='--', linewidth=0.5)
    
    # Sauvegarde du graphique à barres
    plt.savefig(f'{folder_name}/bar_{column}.png', bbox_inches='tight', facecolor='white')
    plt.close()


In [ ]:
# Identifier les colonnes avec des valeurs constantes
cols_to_drop = [col for col in general_data_copy.columns if general_data_copy[col].nunique() == 1]
# Supprimer ces colonnes du DataFrame
general_data_copy.drop(cols_to_drop, axis=1, inplace=True)
# Afficher les colonnes à supprimer
print(f"Colonnes à supprimer car elles contiennent des valeurs constantes : {cols_to_drop}")

In [ ]:
imputer = SimpleImputer(strategy='mean')

cols_to_impute = ['NumCompaniesWorked', 'TotalWorkingYears', 'EnvironmentSatisfaction', 'JobSatisfaction', 'WorkLifeBalance']
general_data_copy[cols_to_impute] = imputer.fit_transform(general_data_copy[cols_to_impute])

In [ ]:
general_data_copy.info()

In [ ]:
#Suppresion des données pour raison éthique
general_data_copy.drop(['EmployeeID', 'Age', 'Gender', 'MaritalStatus'], axis=1, inplace=True)


In [ ]:
#On cherche la corrélation entre les données

import seaborn as sns
sns.set()

corr_cols = general_data_copy
corr = corr_cols.corr()
plt.figure(figsize=(16,10))
sns.heatmap(corr,annot=True)

# Création du dossier pour les graphiques s'il n'existe pas
folder_name = 'analyse'
os.makedirs(folder_name, exist_ok=True)
plt.savefig(f'{folder_name}/correlation_matrix.png', bbox_inches='tight', facecolor='white')

plt.show()

In [ ]:
sns.pairplot(general_data_copy,diag_kind='auto')
plt.show()

In [ ]:
# Can you check to see if there are any missing values in our data set
general_data_copy.shape

In [ ]:
#Calcul du taux de turnover, basé sur la colonne Attrition
turnover_rate = general_data_copy.Attrition.value_counts() / 4410
turnover_rate

In [ ]:
general_data_copy.describe()

In [ ]:
#On affiche la répartition des employés sur chaque attribut en fonction de "Attrition"
turnover_Summary = general_data_copy.groupby('Attrition')
turnover_Summary.mean()

In [ ]:
turnover_Summary.std()

In [ ]:
corr = general_data_copy.corr()
sns.heatmap(corr, 
            xticklabels=corr.columns.values,
            yticklabels=corr.columns.values)
plt.title('Heatmap of Correlation Matrix')
corr

In [ ]:
sns.lmplot(x='YearsWithCurrManager', y='YearsSinceLastPromotion', data=general_data_copy,
           fit_reg=False, # No regression line
           hue='Attrition')   # Color by evolution stage